presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,103 @@ private PinotExpression handleLogicalBinary(
+            }
             return derived(format(
-                    "(%s %s %s)",
-                    arguments.get(0).accept(this, context).getDefinition(),
-                    operator,
-                    arguments.get(1).accept(this, context).getDefinition()));
+                "(%s %s %s)",
+                arguments.get(0).accept(this, context).getDefinition(),
+                operator,
+                arguments.get(1).accept(this, context).getDefinition()));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private PinotExpression handleDateOrTimestampBinaryExpression(String operator, List<RowExpression> arguments, Function<VariableReferenceExpression, Selection> context)
+    {
+        String left = arguments.get(0).accept(this, context).getDefinition();
+        String right = arguments.get(1).accept(this, context).getDefinition();
+        left = handleTimeStampAndDateCast(arguments.get(1), left);
+        right = handleTimeStampAndDateCast(arguments.get(0), right);
+        return derived(format("(%s %s %s)", left, operator, right));
+    }
+
+    private boolean isDateOrTimestampExpression(List<RowExpression> arguments)
+    {
+        for (RowExpression rowExpression : arguments) {
+            if (!(rowExpression instanceof CallExpression)) {
+                continue;
+            }
+            CallExpression callExpr = (CallExpression) rowExpression;
+            if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+                continue;
+            }
+            if (callExpr.getArguments().size() != 1) {
+                continue;
+            }
+            Type inputType = callExpr.getArguments().get(0).getType();
+            Type expectedType = callExpr.getType();
+            if (inputType == DateType.DATE && (expectedType == TimestampType.TIMESTAMP || expectedType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
+                return true;
+            }
+            if ((inputType == TimestampType.TIMESTAMP || inputType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE) && expectedType == DateType.DATE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String handleTimeStampAndDateCast(RowExpression rowExpression, String timeStampOrDateString)
+    {
+        if (!(rowExpression instanceof CallExpression)) {
+            return timeStampOrDateString;
+        }
+        CallExpression callExpr = (CallExpression) rowExpression;
+        if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+            return timeStampOrDateString;
+        }
+        if (callExpr.getArguments().size() != 1) {
+            return timeStampOrDateString;
+        }
+
+        // Handle the binary comparison logic of <DATE/TIMESTAMP field> <binary ops> <DATE/TIMESTAMP literal>.
```

Comment:
Should this comment be moved to handleDateOrTimestampBinaryExpression since it has a reference to the "binary op" ?